tests/integration/shutdown.tcl

```diff
@@ -0,0 +1,157 @@
+foreach how {sigterm shutdown} {
+    test "Shutting down master waits for replica to catch up ($how)" {
+        start_server {} {
+            start_server {} {
+                set master [srv -1 client]
+                set master_host [srv -1 host]
+                set master_port [srv -1 port]
+                set master_pid [srv -1 pid]
+                set replica [srv 0 client]
+                set replica_pid [srv 0 pid]
+
+                # Config master.
+                $master config set repl-diskless-sync yes
+                $master config set repl-diskless-sync-delay 1
+                $master config set shutdown-timeout 300 ; # 5min for slow CI
+
+                # Config replica.
+                $replica replicaof $master_host $master_port
+                $replica config set repl-diskless-load swapdb
+                wait_for_sync $replica
+
+                # Preparation: Set k to 1 on both master and replica.
+                $master set k 1
+                wait_for_ofs_sync $master $replica
+
+                # Pause the replica.
+                exec kill -SIGSTOP $replica_pid
+                after 10
+
+                # Fill upp the replication socket buffers.
+                set junk_size 10000
+                for {set i 0} {$i < $junk_size} {incr i} {
+                    $master set "key.$i.junkjunkjunk" \
+                        "value.$i.blablablablablablablablabla"
+                }
+
+                # Incr k and immediately shutdown master.
+                $master incr k
+                switch $how {
+                    sigterm {
+                        exec kill -SIGTERM $master_pid
+                    }
+                    shutdown {
+                        set rd [redis_deferring_client -1]
+                        $rd shutdown
+                    }
+                }
+
+                # Wake up replica and check if master has waited for it.
+                after 1000
```

Comment:
i'd rather not have an inherent delay of 1 second.
maybe instead we can check the master is still responding (GET command and INFO command)
maybe in addition to that, set the cron hz to a high number, and have an `after` of 2 cron intervals?

Reply:
To check that it responds to INFO and GET isn't the same thing as checking that it actually waits for the replica offset, but it's a pretty good indication. If we instead rely on the new INFO shutdown lines that I'll add, then we don't need this delay.

Reply:
i think that if we know that the shutdown thing is attempted at cron, and we set the hz to be fast and wait for a few cron triggers, it's enough.
when you think of it, that 1 second delay isn't a guarantee of anything either... why 1 and not 5?